### PR TITLE
``pytest.warns`` multiple argument handling

### LINKED ISSUE #11906: Bug in 8.0.0 `pytest.warns` handling
There seems to be a handling error in 8.0.0, when `pytest.warns` fails to match the message. **Seems that the warning class itself is being called towards the end of the handling...**

```python
import pytest
import warnings

assert pytest.__version__ == "8.0.0"

class AWarning(UserWarning):
    
    def __init__(self, a, b):
        pass

a, b = 1, 2
with pytest.warns(AWarning, match="not gonna match"):
    warnings.warn(AWarning(a, b))
```
```
---------------------------------------------------------------------------
Failed                                    Traceback (most recent call last)
    [... skipping hidden 1 frame]

File ~\...\market_prices\.venv\lib\site-packages\_pytest\outcomes.py:187, in fail(reason, pytrace, msg)
    186 reason = _resolve_msg_to_reason("fail", reason, msg)
--> 187 raise Failed(msg=reason, pytrace=pytrace)

Failed: DID NOT WARN. No warnings of type (<class '__main__.AWarning'>,) matching the regex were emitted.
 Regex: not gonna match
 Emitted warnings: [AWarning(1, 2)].

During handling of the above exception, another exception occurred:

TypeError                                 Traceback (most recent call last)
Cell In[1], line 13
     11 a, b = 1, 2
     12 with pytest.warns(AWarning, match="not gonna match"):
---> 13     warnings.warn(AWarning(a, b))

File ~\...\market_prices\.venv\lib\site-packages\_pytest\recwarn.py:333, in WarningsChecker.__exit__(self, exc_type, exc_val, exc_tb)
    331 for w in self:
    332     if not self.matches(w):
--> 333         warnings.warn_explicit(
    334             str(w.message),
    335             w.message.__class__,  # type: ignore[arg-type]
    336             w.filename,
    337             w.lineno,
    338             module=w.__module__,
    339             source=w.source,
    340         )

TypeError: __init__() missing 1 required positional argument: 'b'
```

**NB: only seems to happen if the Warning class takes more than one argument.** For example, if arg b is stripped from all the above then the failure to match is handled as expected.

Also, worked fine in 7.4.4:

```python
import pytest
import warnings

assert pytest.__version__ == "7.4.4"

class AWarning(UserWarning):
    
    def __init__(self, a, b):
        pass

a, b = 1, 2
with pytest.warns(AWarning, match="not gonna match"):
    warnings.warn(AWarning(a, b))
```
```
---------------------------------------------------------------------------
Failed                                    Traceback (most recent call last)
Cell In[1], line 13
     11 a, b = 1, 2
     12 with pytest.warns(AWarning, match="not gonna match"):
---> 13     warnings.warn(AWarning(a, b))

    [... skipping hidden 1 frame]

File ~\...\market_prices\.venv\lib\site-packages\_pytest\outcomes.py:198, in fail(reason, pytrace, msg)
    196 __tracebackhide__ = True
    197 reason = _resolve_msg_to_reason("fail", reason, msg)
--> 198 raise Failed(msg=reason, pytrace=pytrace)

Failed: DID NOT WARN. No warnings of type (<class '__main__.AWarning'>,) matching the regex were emitted.
 Regex: not gonna match
 Emitted warnings: [AWarning(1, 2)]
```

Thanks is advance for looking at this!

OS: 'Windows-10-10.0.22631-SP0'

### PR DESCRIPTION
**Pull request by reaganjlee**
Fixes #11906

### CODE REVIEW

File: src/_pytest/recwarn.py [hunk 1]
```diff
@@ -322,10 +322,10 @@ def found_str():
             for w in self:
                 if not self.matches(w):
                     warnings.warn_explicit(
-                        str(w.message),
-                        w.message.__class__,  # type: ignore[arg-type]
-                        w.filename,
-                        w.lineno,
+                        message=w.message,  # type: ignore[arg-type]
```
Comment (bluetech) [inline src/_pytest/recwarn.py:325]: The docs say:

> message must be a string and category a subclass of [Warning](https://docs.python.org/3/library/exceptions.html#Warning) or message may be a [Warning](https://docs.python.org/3/library/exceptions.html#Warning) instance, in which case category will be ignored.

So I think this should be:

```suggestion
                        message=w,
```

No?

**Review by bluetech — CHANGES_REQUESTED** [review]

File: src/_pytest/recwarn.py (hunk 1 above)
Comment (reaganjlee) [reply src/_pytest/recwarn.py:325]: Yes that makes sense!

Comment (reaganjlee) [reply src/_pytest/recwarn.py:325]: Hm, I'm having a bit of trouble with the mypy and the requirements for `warn_explicit()` (without just type ignoring or casting it). Would you mind taking a look?

Comment (bluetech) [reply src/_pytest/recwarn.py:325]: After trying it and reading the `warn_explicit` code, I think we should be good with just this:

```
                        message=w.message,
                        category=w.category,
```

File: testing/test_recwarn.py [hunk 2]
```diff
@@ -477,3 +477,14 @@ def test_catch_warning_within_raise(self) -> None:
             with pytest.raises(ValueError, match="some exception"):
                 warnings.warn("some warning", category=FutureWarning)
                 raise ValueError("some exception")
+
+    def test_multiple_arg_custom_warning(self) -> None:
```
Comment (nicoddemus) [inline testing/test_recwarn.py:481]: Let's just add a docstring to mention the original issue for future reference (I know we can eventually find it via git-blame, but often things move around, code is refactored/reformatted, so I think it pays to just slap the issue number along with the test):

```suggestion
    def test_multiple_arg_custom_warning(self) -> None:
        """pytest.warns with custom warning subclass with multiple arguments (#11906)."""
```

**Review by nicoddemus — COMMENTED** [review]
LGTM!

Left a small suggest, other than that we also need a changelog, please create `changelog/11906.bugfix.rst` with something along these lines:

```rst
Fix regression using :func:`pytest.warns` with custom warning subclasses which have more than one parameter in their `__init__`.
```

File: src/_pytest/recwarn.py [hunk 3]
```diff
@@ -322,10 +322,9 @@ def found_str():
             for w in self:
                 if not self.matches(w):
                     warnings.warn_explicit(
-                        str(w.message),
-                        w.message.__class__,  # type: ignore[arg-type]
-                        w.filename,
-                        w.lineno,
-                        module=w.__module__,
```
Comment (bluetech) [inline src/_pytest/recwarn.py:329]: Let's keep the `module` in the PR as-is, the fix for that should go in a separate PR (it's a separate issue).

**Review by bluetech — APPROVED** [review]
Thanks @reaganjlee!

**Comment by bluetech** [comment]
@reaganjlee There is a conflict, can you rebase on latest main please?

File: src/_pytest/recwarn.py [hunk 4]
```diff
@@ -336,7 +332,9 @@ def found_str():
 
     @staticmethod
     def _validate_message(wrn: Any) -> None:
-        if not isinstance(msg := wrn.message.args[0], str):
```
Comment (bluetech) [inline src/_pytest/recwarn.py:339]: Can you explain these changes?

**Review by bluetech — CHANGES_REQUESTED** [review]
Left a question on the new change. Also there are some formatting changes which seem unnecessary so linting fails, can you remove them?

File: src/_pytest/recwarn.py (hunk 4 above)
Comment (reaganjlee) [reply src/_pytest/recwarn.py:339]: After making the other changes, the code failed some tests from #11804 that solved #10865, where the incorrect arguments then defaulted to becoming `UserWarning`s. However, the implementation also caught the custom `CustomWarning` class (where `wrn.message.args[0] = int` from `CustomWarning`'s `__init__`, which this looked to fix.
This looks related to #11954

Comment (reaganjlee) [reply src/_pytest/recwarn.py:339]: Hm, this looks like it could be its own PR potentially. Another implementation: #11959

**Comment by bluetech** [comment]
Merged through #11991, thanks @reaganjlee!

**Pull request closed without merging**